Release/src/http/client/http_client_curl.cpp

```diff
@@ -0,0 +1,2379 @@
+    {
+        assert(block != nullptr);
+        m_externalData = block;
+        m_size = length;
+    }
+
+    inline bool is_internally_allocated() const { return m_externalData == nullptr; }
+
+    inline uint8_t* get() { return is_internally_allocated() ? &m_internalData[0] : m_externalData; }
+
+    inline size_t size() const { return is_internally_allocated() ? m_internalData.size() : m_size; }
+};
+
+// Possible ways a message body can be sent/received.
+enum msg_body_type
+{
+    no_body,
+    content_length_chunked,
+    transfer_encoding_chunked
+};
+
+// Additional information necessary to track a WinHTTP request.
+class winhttp_request_context final : public request_context
+{
+public:
+    // Factory function to create requests on the heap.
+    static std::shared_ptr<request_context> create_request_context(
+        const std::shared_ptr<_http_client_communicator>& client, const http_request& request)
+    {
+        std::shared_ptr<winhttp_request_context> ret(new winhttp_request_context(client, request));
+        ret->m_self_reference = ret;
+        return std::move(ret);
+    }
+
+    ~winhttp_request_context() { cleanup(); }
+
+    void allocate_request_space(_In_opt_ uint8_t* block, size_t length)
+    {
+        if (block == nullptr)
+            m_body_data.allocate_space(length);
+        else
+            m_body_data.reassign_to(block, length);
+    }
+
+    void allocate_reply_space(_In_opt_ uint8_t* block, size_t length)
+    {
+        if (block == nullptr)
+            m_body_data.allocate_space(length);
+        else
+            m_body_data.reassign_to(block, length);
+    }
+
+    bool is_externally_allocated() const { return !m_body_data.is_internally_allocated(); }
+
+    HINTERNET m_request_handle;
+    std::weak_ptr<winhttp_request_context>*
+        m_request_handle_context; // owned by m_request_handle to be delete'd by WINHTTP_CALLBACK_STATUS_HANDLE_CLOSING
+
+    bool m_proxy_authentication_tried;
+    bool m_server_authentication_tried;
+
+    msg_body_type m_bodyType;
+
+    utility::size64_t m_remaining_to_write;
+
+    std::char_traits<uint8_t>::pos_type m_startingPosition;
+
+    // If the user specified that to guarantee data buffering of request data, in case of challenged authentication
+    // requests, etc... Then if the request stream buffer doesn't support seeking we need to copy the body chunks as it
+    // is sent.
+    concurrency::streams::istream m_readStream;
+    std::unique_ptr<concurrency::streams::container_buffer<std::vector<uint8_t>>> m_readBufferCopy;
+    virtual concurrency::streams::streambuf<uint8_t> _get_readbuffer() { return m_readStream.streambuf(); }
+
+    // This self reference will keep us alive until finish() is called.
+    std::shared_ptr<winhttp_request_context> m_self_reference;
+    memory_holder m_body_data;
+
+    // Compress/decompress-related processing state lives here
+    class compression_state
```

Comment:
Is this copy pasta from somewhere else? If so it should be extracted into a common .cpp instead of duplicated.

Reply:
merged to http_client_winhttp.cpp